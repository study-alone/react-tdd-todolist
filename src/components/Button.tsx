import type { FC } from 'react'
import Styled from 'styled-components'

type ContainerProps = {
	readonly backgroundColor: string
	readonly hoverColor: string
}
const Contanier = Styled.div<ContainerProps>`
    text-align: center;
    background-color: ${(props) => props.backgroundColor};
    padding: 10px 20px;
    border-radius: 8px;
    cursor: pointer;
    &:hover {
        background-color: ${(props) => props.hoverColor};
    }
    &:active {
        box-shadow: inset 5px 5px 10px rgba(0, 0, 0, 0.2);
    }
`
const Label = Styled.div`
    color: #fff;
    font-size: 16px;
`

type Props = {
	readonly label: string
	readonly backgroundColor?: string
	readonly hoverColor?: string
	readonly onClick?: () => void
}
export const Button: FC<Props> = ({
	label,
	backgroundColor = '#304ffe',
	hoverColor = '#1e40ff',
	onClick,
}) => {
	return (
		<Contanier {...{ hoverColor, backgroundColor }} onClick={onClick}>
			<Label>{label}</Label>
		</Contanier>
	)
}
