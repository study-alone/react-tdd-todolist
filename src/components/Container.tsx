import Styled from 'styled-components'

export const Container = Styled.div`
    min-height: 100vh;
    display: flex;
    align-items: center;
    justify-content: center;
    flex-direction: column;
`
