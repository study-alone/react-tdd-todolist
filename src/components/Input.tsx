import type { ChangeEventHandler, FC } from 'react'
import Styled from 'styled-components'

const InputBox = Styled.input`
    font-size: 16px;
    padding: 10px;
    border-radius: 8px;
    border: 1px solid #bdbdbd;
    outline: none;
`

type Props = {
	readonly placeholder?: string
	readonly onChange?: (text: string) => void
	readonly value?: string
}

export const Input: FC<Props> = ({
	placeholder = '할 일 입력',
	value,
	onChange,
}) => {
	const change: ChangeEventHandler<HTMLInputElement> = (event) => {
		typeof onChange === 'function' && onChange(event.target.value)
	}
	return (
		<InputBox placeholder={placeholder} value={value} onChange={change} />
	)
}
