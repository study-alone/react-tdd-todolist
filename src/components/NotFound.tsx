import Styled from 'styled-components'
import type { FC } from 'react'

const Container = Styled.div`
    font-size: 20px;
`

export const NotFound: FC = () => {
	return <Container>Not Found</Container>
}
