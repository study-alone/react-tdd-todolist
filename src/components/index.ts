export * from '@/components/Button'
export * from '@/components/Container'
export * from '@/components/Input'
export * from '@/components/TodoItem'
export * from '@/components/InputContainer'
export * from '@/components/TodoList'
export * from '@/components/PageHeader'
export * from '@/components/NotFound'
