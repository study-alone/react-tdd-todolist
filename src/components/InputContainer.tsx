import { useContext, useState } from 'react'
import type { FC } from 'react'
import Styled from 'styled-components'
import { Button, Input } from '@/components'
import { TodoListContext } from '@/context'

const Container = Styled.div`
    display: flex;
`

type Props = {
	// readonly todo?: string
	// readonly onChange?: (text: string) => void
	readonly onAdd?: () => void
}

export const InputContainer: FC<Props> = ({ onAdd }) => {
	const [todo, setTodo] = useState('')
	const { addTodo } = useContext(TodoListContext)

	const onAddTodo = () => {
		addTodo(todo)
		setTodo('')
		if (todo && typeof onAdd === 'function') {
			onAdd()
		}
	}

	return (
		<Container>
			<Input
				placeholder="할 일을 입력해주세요."
				value={todo}
				onChange={setTodo}
			/>
			<Button label="추가" onClick={onAddTodo} />
		</Container>
	)
}
