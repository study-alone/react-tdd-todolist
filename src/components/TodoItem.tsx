import type { FC } from 'react'
import { Link } from 'react-router-dom'
import Styled from 'styled-components'
import { Button } from '@/components'

const Container = Styled.div`
    display: flex;
    border-bottom: 1px solid #bdbdbd;
    align-items: center;
    margin: 10px;
    padding: 10px;
`
const Label = Styled(Link)`
    flex: 1;
    font-size: 16px;
    margin-right: 20px;
    text-decoration: none;
`

type Props = {
	readonly id: number
	readonly label: string
	readonly onDelete?: () => void
}
export const TodoItem: FC<Props> = ({ label, onDelete, id }) => {
	return (
		<Container>
			<Label to={`/detail/${id}`}>{label}</Label>
			<Button
				label="삭제"
				backgroundColor="#ff1744"
				hoverColor="#f01440"
				onClick={onDelete}
			/>
		</Container>
	)
}
