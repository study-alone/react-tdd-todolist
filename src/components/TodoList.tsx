import { useContext } from 'react'
import Styled from 'styled-components'
import { TodoItem } from '@/components'
import type { FC } from 'react'
import { TodoListContext } from '@/context'

const TodoListContainer = Styled.div`
    min-width: 350px;
    height: 400px;
    overflow-y: scroll;
    border: 1px solid #bdbdbd;
    margin-bottom: 20px;
`

type Props = {
	// readonly todoList?: string[]
	// readonly deleteTodo?: (index: number) => void
}

export const TodoList: FC<Props> = ({}) => {
	const { todoList, deleteTodo } = useContext(TodoListContext)

	return (
		<TodoListContainer data-testid="todoList">
			{todoList.map((item, index) => (
				<TodoItem
					key={item}
					id={index}
					label={item}
					onDelete={() => deleteTodo(index)}
				/>
			))}
		</TodoListContainer>
	)
}
