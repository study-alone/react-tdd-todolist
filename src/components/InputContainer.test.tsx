import React from 'react'
import { screen, render, fireEvent } from '@testing-library/react'
import 'jest-styled-components'
import { InputContainer } from '@/components'
import { TodoListProvider } from '@/context'

const LS_KEY = 'todoList'

describe('<InputContainer />', () => {
	it('컴포넌트가 잘 렌더링 되는가?', () => {
		const { container } = render(<InputContainer />)

		const input = screen.getByPlaceholderText('할 일을 입력해주세요.')
		expect(input).toBeInTheDocument()

		const button = screen.getByText('추가')
		expect(button).toBeInTheDocument()

		expect(container).toMatchSnapshot()
	})

	it('데이터 추가 후 input value가 빈 값이 되는가?', () => {
		render(<InputContainer />)

		const input =
			screen.getByPlaceholderText<HTMLInputElement>(
				'할 일을 입력해주세요.',
			)
		const button = screen.getByText('추가')

		expect(input.value).toBe('')
		fireEvent.change(input, { target: { value: 'study react 1' } })
		expect(input.value).toBe('study react 1')
		fireEvent.click(button)
		expect(input.value).toBe('')
	})

	it('Context를 통한 localstorage 추가가 잘 되는가?', () => {
		render(
			<TodoListProvider>
				<InputContainer />
			</TodoListProvider>,
		)

		const input = screen.getByPlaceholderText('할 일을 입력해주세요.')
		const button = screen.getByText('추가')

		expect(localStorage.getItem(LS_KEY)).toBeNull()

		fireEvent.change(input, { target: { value: 'study react 1' } })
		fireEvent.click(button)

		expect(localStorage.getItem(LS_KEY)).toBe(
			JSON.stringify(['study react 1']),
		)
	})
})
