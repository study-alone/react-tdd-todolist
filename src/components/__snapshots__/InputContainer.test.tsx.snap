// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<InputContainer /> 컴포넌트가 잘 렌더링 되는가? 1`] = `
.c2 {
  text-align: center;
  background-color: #304ffe;
  padding: 10px 20px;
  border-radius: 8px;
  cursor: pointer;
}

.c2:hover {
  background-color: #1e40ff;
}

.c2:active {
  box-shadow: inset 5px 5px 10px rgba(0,0,0,0.2);
}

.c3 {
  color: #fff;
  font-size: 16px;
}

.c1 {
  font-size: 16px;
  padding: 10px;
  border-radius: 8px;
  border: 1px solid #bdbdbd;
  outline: none;
}

.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
}

<div>
  <div
    class="c0"
  >
    <input
      class="c1"
      placeholder="할 일을 입력해주세요."
      value=""
    />
    <div
      class="c2"
    >
      <div
        class="c3"
      >
        추가
      </div>
    </div>
  </div>
</div>
`;
